<img src="https://raw.githubusercontent.com/alan-barzilay/NLPortugues/master/imagens/logo_nlportugues.png"  style="height:65%" align="right">


# Lista 2 - NN & Backpropagation
**Nome: Felipe de Lima Peressim**

**Numero Usp: 11823558**


_________________________________________

Antes de começar o exercício,  não se esqueça de instalar todos os pacotes necessários para a sua execução.  

In [1]:
import tensorflow as tf
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
from tensorflow import keras
from sklearn.model_selection import train_test_split
%matplotlib inline
tf.__version__

/home/felipe/.local/lib/python3.6/site-packages/statsmodels/tools/_testing.py:19: FutureWarning: pandas.util.testing is deprecated. Use the functions in the public API at pandas.testing instead.
  import pandas.util.testing as tm


'2.2.0'

## Perceptron

O perceptron é uma "rede neural" de um só neurônio.  No nosso caso, temos a rede mais simpes possível, com uma só entrada e uma só saída, sem ativação.

Temos 100 dados que serão usados para treinar 300 épocas do percéptron.

Vamos utilizar o modelo perceptron para aprender uma simples regressão linear, o objetivo é faze-lo aprender uma simples equação linear e tambem se acostumar com a sintaxe e funcionamento do TensorFlow


In [49]:
def f1(x):
    '''
    Funcao a ser aprendida
    '''
    return 5 + 10*x

In [50]:
xs = np.linspace(0,10,100)  # gera 100 valores no intervalo [0.10]
ys = f1(xs)                 # computa o valor de f1 nestes 100 valores
print(len(xs), "xs=", xs)
print(len(ys), "ys=", ys)


100 xs= [ 0.          0.1010101   0.2020202   0.3030303   0.4040404   0.50505051
  0.60606061  0.70707071  0.80808081  0.90909091  1.01010101  1.11111111
  1.21212121  1.31313131  1.41414141  1.51515152  1.61616162  1.71717172
  1.81818182  1.91919192  2.02020202  2.12121212  2.22222222  2.32323232
  2.42424242  2.52525253  2.62626263  2.72727273  2.82828283  2.92929293
  3.03030303  3.13131313  3.23232323  3.33333333  3.43434343  3.53535354
  3.63636364  3.73737374  3.83838384  3.93939394  4.04040404  4.14141414
  4.24242424  4.34343434  4.44444444  4.54545455  4.64646465  4.74747475
  4.84848485  4.94949495  5.05050505  5.15151515  5.25252525  5.35353535
  5.45454545  5.55555556  5.65656566  5.75757576  5.85858586  5.95959596
  6.06060606  6.16161616  6.26262626  6.36363636  6.46464646  6.56565657
  6.66666667  6.76767677  6.86868687  6.96969697  7.07070707  7.17171717
  7.27272727  7.37373737  7.47474747  7.57575758  7.67676768  7.77777778
  7.87878788  7.97979798  8.08080808  8.181

In [4]:
#Definindo, compilando e treinando nosso modelo
model = tf.keras.Sequential([
    keras.Input(shape=(1,)),
    keras.layers.Dense(units=1),
])

model.compile(optimizer="sgd", loss="mean_squared_error")
model.fit(xs,ys,epochs=300)

Epoch 1/300
4/4 [==============================] - 0s 5ms/step - loss: 1418.0372
Epoch 2/300
4/4 [==============================] - 0s 3ms/step - loss: 2.8900
Epoch 3/300
4/4 [==============================] - 0s 2ms/step - loss: 2.8248
Epoch 4/300
4/4 [==============================] - 0s 1ms/step - loss: 2.6407
Epoch 5/300
4/4 [==============================] - 0s 2ms/step - loss: 2.5853
Epoch 6/300
4/4 [==============================] - 0s 3ms/step - loss: 2.3974
Epoch 7/300
4/4 [==============================] - 0s 3ms/step - loss: 2.3016
Epoch 8/300
4/4 [==============================] - 0s 2ms/step - loss: 2.2957
Epoch 9/300
4/4 [==============================] - 0s 3ms/step - loss: 2.2120
Epoch 10/300
4/4 [==============================] - 0s 2ms/step - loss: 2.1516
Epoch 11/300
4/4 [==============================] - 0s 2ms/step - loss: 1.9735
Epoch 12/300
4/4 [==============================] - 0s 2ms/step - loss: 1.8516
Epoch 13/300
4/4 [==============================] - 0s 2ms

4/4 [==============================] - 0s 2ms/step - loss: 8.7334e-04
Epoch 207/300
4/4 [==============================] - 0s 1ms/step - loss: 8.2909e-04
Epoch 208/300
4/4 [==============================] - 0s 2ms/step - loss: 8.0360e-04
Epoch 209/300
4/4 [==============================] - 0s 2ms/step - loss: 7.4268e-04
Epoch 210/300
4/4 [==============================] - 0s 2ms/step - loss: 7.1674e-04
Epoch 211/300
4/4 [==============================] - 0s 2ms/step - loss: 7.0549e-04
Epoch 212/300
4/4 [==============================] - 0s 1ms/step - loss: 6.5414e-04
Epoch 213/300
4/4 [==============================] - 0s 2ms/step - loss: 6.2399e-04
Epoch 214/300
4/4 [==============================] - 0s 2ms/step - loss: 6.1926e-04
Epoch 215/300
4/4 [==============================] - 0s 2ms/step - loss: 5.7828e-04
Epoch 216/300
4/4 [==============================] - 0s 2ms/step - loss: 5.9337e-04
Epoch 217/300
4/4 [==============================] - 0s 2ms/step - loss: 5.4302e-04
Epoch 

In [5]:
model.summary()

Model: "sequential"
_________________________________________________________________
Layer (type)                 Output Shape              Param #   
dense (Dense)                (None, 1)                 2         
Total params: 2
Trainable params: 2
Non-trainable params: 0
_________________________________________________________________


In [6]:
print("prediction: "+ str(model.predict([17]))+"      real value: " + str(f1(17)))

prediction: [[175.01234]]      real value: 175


In [7]:
val = np.linspace(0,10,63)
model.evaluate(x=val, y=f1(val))

2/2 [==============================] - 0s 686us/step - loss: 2.0905e-05


2.0905352357658558e-05

A função `evaluate` retorna o "custo" (loss) da avaliação, definido na compilação.  Nesse caso, o valor reportado é o erro quadrático médio (MSE). 

In [51]:
#Definindo, compilando e treinando nosso modelo
model = tf.keras.Sequential([
    keras.Input(shape=(1,)),
    keras.layers.Dense(units=1),
])

model.compile(optimizer="adam", loss="mean_squared_error")
model.fit(xs,ys,epochs=300)

Epoch 1/300
4/4 [==============================] - 0s 763us/step - loss: 3017.0020
Epoch 2/300
4/4 [==============================] - 0s 719us/step - loss: 3014.0359
Epoch 3/300
4/4 [==============================] - 0s 760us/step - loss: 3011.2471
Epoch 4/300
4/4 [==============================] - 0s 2ms/step - loss: 3008.2859
Epoch 5/300
4/4 [==============================] - 0s 1ms/step - loss: 3005.3882
Epoch 6/300
4/4 [==============================] - 0s 3ms/step - loss: 3002.5034
Epoch 7/300
4/4 [==============================] - 0s 2ms/step - loss: 2999.6912
Epoch 8/300
4/4 [==============================] - 0s 2ms/step - loss: 2996.8560
Epoch 9/300
4/4 [==============================] - 0s 1ms/step - loss: 2994.1384
Epoch 10/300
4/4 [==============================] - 0s 4ms/step - loss: 2991.1963
Epoch 11/300
4/4 [==============================] - 0s 3ms/step - loss: 2988.3821
Epoch 12/300
4/4 [==============================] - 0s 1ms/step - loss: 2985.6306
Epoch 13/300
4/4 [=

4/4 [==============================] - 0s 2ms/step - loss: 2488.7117
Epoch 199/300
4/4 [==============================] - 0s 2ms/step - loss: 2486.1470
Epoch 200/300
4/4 [==============================] - 0s 2ms/step - loss: 2483.6501
Epoch 201/300
4/4 [==============================] - 0s 2ms/step - loss: 2481.1609
Epoch 202/300
4/4 [==============================] - 0s 2ms/step - loss: 2478.6672
Epoch 203/300
4/4 [==============================] - 0s 3ms/step - loss: 2476.2859
Epoch 204/300
4/4 [==============================] - 0s 3ms/step - loss: 2473.7446
Epoch 205/300
4/4 [==============================] - 0s 1ms/step - loss: 2471.3645
Epoch 206/300
4/4 [==============================] - 0s 1ms/step - loss: 2468.8970
Epoch 207/300
4/4 [==============================] - 0s 2ms/step - loss: 2466.3533
Epoch 208/300
4/4 [==============================] - 0s 2ms/step - loss: 2463.9561
Epoch 209/300
4/4 [==============================] - 0s 2ms/step - loss: 2461.5579
Epoch 210/300
4/4 

## Aprendendo uma função não linear

In [35]:
def f2(x):
    '''
    Funcao não linear a ser aprendida
    '''
    return (x**2 + x*3 + 4)/200

In [36]:
x = np.linspace(0,10,100)
y = f2(x)

# <font color='blue'>Questão 1 </font>
Defina as camadas para esta rede neural e treine seu modelo, note que a saída unitária _não deve_ ter função de ativação (por que?).

**Por tratar de um problema de regressão, nós queremos ter como resposta um valor $y$ que pertença à imagem da função que queiramos prever, ou seja, normalmente um valor $y$ tal que $y \in {\rm I\!R}$. Normalmente o intervalo de valores da distribuição dos dados se encontra em um intervalo diferente em que os resultados produzidos pelas funções de ativação se encontram.**

**Se  camada de saída tivesse uma função de ativação diferente da função identidade $(f(x) = x)$, o resultado produzido, seria mapeado para um intervalo de valores restrito, que possivelmente não abrangeria o mesmo intervalo de valores da função alvo e portanto produziria resultados inconsistentes.**

**O uso de uma função de ativação, outra além da função identidade  na camada de saída da rede, só seria plausível se a distribuição dos dados estivesse retrita a mesma faixa de valores da função de ativação em questão.**

In [37]:
model = tf.keras.Sequential([    
    keras.Input(shape=(1,)),
    keras.layers.Dense(units = 10, activation = 'relu'),
    keras.layers.Dense(units = 10, activation = 'relu'),
    keras.layers.Dense(units = 10, activation = 'relu'),
    keras.layers.Dense(units = 1)
])
model.compile(optimizer="adam", loss="mean_squared_error")
model.fit(x,y,epochs=400)

Epoch 1/400
4/4 [==============================] - 0s 753us/step - loss: 0.0107
Epoch 2/400
4/4 [==============================] - 0s 2ms/step - loss: 0.0033
Epoch 3/400
4/4 [==============================] - 0s 923us/step - loss: 0.0049
Epoch 4/400
4/4 [==============================] - 0s 2ms/step - loss: 0.0035
Epoch 5/400
4/4 [==============================] - 0s 2ms/step - loss: 0.0028
Epoch 6/400
4/4 [==============================] - 0s 2ms/step - loss: 0.0029
Epoch 7/400
4/4 [==============================] - 0s 3ms/step - loss: 0.0022
Epoch 8/400
4/4 [==============================] - 0s 2ms/step - loss: 0.0021
Epoch 9/400
4/4 [==============================] - 0s 3ms/step - loss: 0.0022
Epoch 10/400
4/4 [==============================] - 0s 2ms/step - loss: 0.0018
Epoch 11/400
4/4 [==============================] - 0s 2ms/step - loss: 0.0016
Epoch 12/400
4/4 [==============================] - 0s 2ms/step - loss: 0.0015
Epoch 13/400
4/4 [==============================] - 0s 2m

4/4 [==============================] - 0s 3ms/step - loss: 8.1472e-05
Epoch 198/400
4/4 [==============================] - 0s 2ms/step - loss: 4.8464e-05
Epoch 199/400
4/4 [==============================] - 0s 3ms/step - loss: 3.9117e-05
Epoch 200/400
4/4 [==============================] - 0s 3ms/step - loss: 2.5615e-05
Epoch 201/400
4/4 [==============================] - 0s 2ms/step - loss: 2.5260e-05
Epoch 202/400
4/4 [==============================] - 0s 2ms/step - loss: 1.7084e-05
Epoch 203/400
4/4 [==============================] - 0s 2ms/step - loss: 1.8296e-05
Epoch 204/400
4/4 [==============================] - 0s 2ms/step - loss: 1.6851e-05
Epoch 205/400
4/4 [==============================] - 0s 2ms/step - loss: 1.6524e-05
Epoch 206/400
4/4 [==============================] - 0s 2ms/step - loss: 1.5945e-05
Epoch 207/400
4/4 [==============================] - 0s 3ms/step - loss: 1.5426e-05
Epoch 208/400
4/4 [==============================] - 0s 3ms/step - loss: 1.6472e-05
Epoch 

4/4 [==============================] - 0s 6ms/step - loss: 1.1168e-05
Epoch 392/400
4/4 [==============================] - 0s 2ms/step - loss: 1.1809e-05
Epoch 393/400
4/4 [==============================] - 0s 2ms/step - loss: 1.0892e-05
Epoch 394/400
4/4 [==============================] - 0s 3ms/step - loss: 5.9162e-06
Epoch 395/400
4/4 [==============================] - 0s 2ms/step - loss: 6.6884e-06
Epoch 396/400
4/4 [==============================] - 0s 2ms/step - loss: 6.6601e-06
Epoch 397/400
4/4 [==============================] - 0s 3ms/step - loss: 8.9890e-06
Epoch 398/400
4/4 [==============================] - 0s 4ms/step - loss: 1.5485e-05
Epoch 399/400
4/4 [==============================] - 0s 2ms/step - loss: 1.7239e-05
Epoch 400/400
4/4 [==============================] - 0s 2ms/step - loss: 1.8616e-05


In [38]:
model.summary()

Model: "sequential_10"
_________________________________________________________________
Layer (type)                 Output Shape              Param #   
dense_30 (Dense)             (None, 10)                20        
_________________________________________________________________
dense_31 (Dense)             (None, 10)                110       
_________________________________________________________________
dense_32 (Dense)             (None, 10)                110       
_________________________________________________________________
dense_33 (Dense)             (None, 1)                 11        
Total params: 251
Trainable params: 251
Non-trainable params: 0
_________________________________________________________________


In [39]:
x_val = np.linspace(0,10,63)
y_val = f2(x_val)

In [40]:
model.evaluate(x=x_val,y=y_val)

2/2 [==============================] - 0s 974us/step - loss: 1.7216e-05


1.7215532352565788e-05

____________________________
# <font color='blue'>Questão 2 </font>



O que acontece se você muda as funções de ativação? Teste algumas diferentes e descreva seus resultados, em especial a tangente hiperbolica



In [25]:
model = tf.keras.Sequential([    
    keras.Input(shape=(1,)),
    keras.layers.Dense(units = 10, activation = 'sigmoid'),
    keras.layers.Dense(units = 10, activation = 'sigmoid'),
    keras.layers.Dense(units = 10, activation = 'sigmoid'),
    keras.layers.Dense(units = 1)
])
model.compile(optimizer="adam", loss="mean_squared_error")
model.fit(x,y,epochs=400)

Epoch 1/400
4/4 [==============================] - 0s 1ms/step - loss: 1.0154
Epoch 2/400
4/4 [==============================] - 0s 2ms/step - loss: 0.9131
Epoch 3/400
4/4 [==============================] - 0s 3ms/step - loss: 0.8171
Epoch 4/400
4/4 [==============================] - 0s 4ms/step - loss: 0.7274
Epoch 5/400
4/4 [==============================] - 0s 3ms/step - loss: 0.6452
Epoch 6/400
4/4 [==============================] - 0s 2ms/step - loss: 0.5688
Epoch 7/400
4/4 [==============================] - 0s 3ms/step - loss: 0.4999
Epoch 8/400
4/4 [==============================] - 0s 3ms/step - loss: 0.4388
Epoch 9/400
4/4 [==============================] - 0s 2ms/step - loss: 0.3830
Epoch 10/400
4/4 [==============================] - 0s 2ms/step - loss: 0.3331
Epoch 11/400
4/4 [==============================] - 0s 3ms/step - loss: 0.2885
Epoch 12/400
4/4 [==============================] - 0s 3ms/step - loss: 0.2501
Epoch 13/400
4/4 [==============================] - 0s 4ms/st

4/4 [==============================] - 0s 2ms/step - loss: 0.0050
Epoch 207/400
4/4 [==============================] - 0s 2ms/step - loss: 0.0049
Epoch 208/400
4/4 [==============================] - 0s 2ms/step - loss: 0.0049
Epoch 209/400
4/4 [==============================] - 0s 3ms/step - loss: 0.0048
Epoch 210/400
4/4 [==============================] - 0s 2ms/step - loss: 0.0048
Epoch 211/400
4/4 [==============================] - 0s 2ms/step - loss: 0.0048
Epoch 212/400
4/4 [==============================] - 0s 2ms/step - loss: 0.0047
Epoch 213/400
4/4 [==============================] - 0s 2ms/step - loss: 0.0046
Epoch 214/400
4/4 [==============================] - 0s 2ms/step - loss: 0.0045
Epoch 215/400
4/4 [==============================] - 0s 2ms/step - loss: 0.0045
Epoch 216/400
4/4 [==============================] - 0s 2ms/step - loss: 0.0044
Epoch 217/400
4/4 [==============================] - 0s 3ms/step - loss: 0.0043
Epoch 218/400
4/4 [==============================] - 0

In [26]:
x_val = np.linspace(0,10,63)
y_val = f2(x_val)

In [27]:
model.evaluate(x=x_val,y=y_val)

2/2 [==============================] - 0s 814us/step - loss: 7.6333e-04


0.0007633315399289131

In [28]:
model = tf.keras.Sequential([    
    keras.Input(shape=(1,)),
    keras.layers.Dense(units = 10, activation = 'tanh'),
    keras.layers.Dense(units = 10, activation = 'tanh'),
    keras.layers.Dense(units = 10, activation = 'tanh'),
    keras.layers.Dense(units = 1)
])
model.compile(optimizer="adam", loss="mean_squared_error")
model.fit(x,y,epochs=400)

Epoch 1/400
4/4 [==============================] - 0s 3ms/step - loss: 0.2985
Epoch 2/400
4/4 [==============================] - 0s 2ms/step - loss: 0.1512
Epoch 3/400
4/4 [==============================] - 0s 2ms/step - loss: 0.0636
Epoch 4/400
4/4 [==============================] - 0s 3ms/step - loss: 0.0270
Epoch 5/400
4/4 [==============================] - 0s 4ms/step - loss: 0.0246
Epoch 6/400
4/4 [==============================] - 0s 3ms/step - loss: 0.0337
Epoch 7/400
4/4 [==============================] - 0s 1ms/step - loss: 0.0364
Epoch 8/400
4/4 [==============================] - 0s 3ms/step - loss: 0.0320
Epoch 9/400
4/4 [==============================] - 0s 3ms/step - loss: 0.0252
Epoch 10/400
4/4 [==============================] - 0s 1ms/step - loss: 0.0210
Epoch 11/400
4/4 [==============================] - 0s 2ms/step - loss: 0.0199
Epoch 12/400
4/4 [==============================] - 0s 4ms/step - loss: 0.0195
Epoch 13/400
4/4 [==============================] - 0s 3ms/st

4/4 [==============================] - 0s 2ms/step - loss: 4.6272e-04
Epoch 204/400
4/4 [==============================] - 0s 2ms/step - loss: 4.5129e-04
Epoch 205/400
4/4 [==============================] - 0s 3ms/step - loss: 5.0042e-04
Epoch 206/400
4/4 [==============================] - 0s 2ms/step - loss: 4.7943e-04
Epoch 207/400
4/4 [==============================] - 0s 2ms/step - loss: 4.5808e-04
Epoch 208/400
4/4 [==============================] - 0s 3ms/step - loss: 4.4722e-04
Epoch 209/400
4/4 [==============================] - 0s 4ms/step - loss: 4.0353e-04
Epoch 210/400
4/4 [==============================] - 0s 3ms/step - loss: 4.0359e-04
Epoch 211/400
4/4 [==============================] - 0s 2ms/step - loss: 4.3826e-04
Epoch 212/400
4/4 [==============================] - 0s 4ms/step - loss: 4.2722e-04
Epoch 213/400
4/4 [==============================] - 0s 4ms/step - loss: 4.8473e-04
Epoch 214/400
4/4 [==============================] - 0s 3ms/step - loss: 4.5695e-04
Epoch 

4/4 [==============================] - 0s 2ms/step - loss: 1.0303e-04
Epoch 398/400
4/4 [==============================] - 0s 2ms/step - loss: 1.0376e-04
Epoch 399/400
4/4 [==============================] - 0s 3ms/step - loss: 1.0738e-04
Epoch 400/400
4/4 [==============================] - 0s 2ms/step - loss: 9.6429e-05


In [29]:
x_val = np.linspace(0,10,63)
y_val = f2(x_val)

In [30]:
model.evaluate(x=x_val,y=y_val)

2/2 [==============================] - 0s 909us/step - loss: 1.0117e-04


0.00010117200872628018

In [31]:
model = tf.keras.Sequential([    
    keras.Input(shape=(1,)),
    keras.layers.Dense(units = 10, activation = 'sigmoid'),
    keras.layers.Dense(units = 10, activation = 'tanh'),
    keras.layers.Dense(units = 10, activation = 'relu'),
    keras.layers.Dense(units = 1)
])
model.compile(optimizer="adam", loss="mean_squared_error")
model.fit(x,y,epochs=400)

Epoch 1/400
4/4 [==============================] - 0s 716us/step - loss: 1.5354
Epoch 2/400
4/4 [==============================] - 0s 824us/step - loss: 1.2690
Epoch 3/400
4/4 [==============================] - 0s 2ms/step - loss: 1.0422
Epoch 4/400
4/4 [==============================] - 0s 2ms/step - loss: 0.8497
Epoch 5/400
4/4 [==============================] - 0s 1ms/step - loss: 0.6911
Epoch 6/400
4/4 [==============================] - 0s 3ms/step - loss: 0.5575
Epoch 7/400
4/4 [==============================] - 0s 4ms/step - loss: 0.4494
Epoch 8/400
4/4 [==============================] - 0s 2ms/step - loss: 0.3622
Epoch 9/400
4/4 [==============================] - 0s 3ms/step - loss: 0.2886
Epoch 10/400
4/4 [==============================] - 0s 2ms/step - loss: 0.2312
Epoch 11/400
4/4 [==============================] - 0s 2ms/step - loss: 0.1834
Epoch 12/400
4/4 [==============================] - 0s 2ms/step - loss: 0.1466
Epoch 13/400
4/4 [==============================] - 0s 2m

4/4 [==============================] - 0s 3ms/step - loss: 0.0021
Epoch 207/400
4/4 [==============================] - 0s 2ms/step - loss: 0.0021
Epoch 208/400
4/4 [==============================] - 0s 3ms/step - loss: 0.0021
Epoch 209/400
4/4 [==============================] - 0s 2ms/step - loss: 0.0020
Epoch 210/400
4/4 [==============================] - 0s 3ms/step - loss: 0.0020
Epoch 211/400
4/4 [==============================] - 0s 2ms/step - loss: 0.0020
Epoch 212/400
4/4 [==============================] - 0s 2ms/step - loss: 0.0020
Epoch 213/400
4/4 [==============================] - 0s 2ms/step - loss: 0.0021
Epoch 214/400
4/4 [==============================] - 0s 3ms/step - loss: 0.0019
Epoch 215/400
4/4 [==============================] - 0s 3ms/step - loss: 0.0020
Epoch 216/400
4/4 [==============================] - 0s 3ms/step - loss: 0.0019
Epoch 217/400
4/4 [==============================] - 0s 3ms/step - loss: 0.0018
Epoch 218/400
4/4 [==============================] - 0

In [32]:
x_val = np.linspace(0,10,63)
y_val = f2(x_val)

In [33]:
model.evaluate(x=x_val,y=y_val)

2/2 [==============================] - 0s 2ms/step - loss: 4.0191e-04


0.00040191086009144783

**<font color='red'> Sua resposta aqui </font>**

**Três funções de ativação foram utilizadas - ReLU, Tanh e Sigmoid - em um modelo de rede neural configurado com 3 camadas ocultas e 10 neurônios em cada camada. Desta forma, um total de quatro redes neurais foram testadas, nas três primeiras todos os neurônios tinham a mesma função de ativação já na quarta e última cada camada contava com uma função de ativação distinta das outras.**

**Através destes testes pode-se observar que ao se experimentar as redes variando as funções de ativação, percebeu-se que houve uma sutis pioras nos resultados exibidos pela função de custo - mse - tanto durante o treinamento quanto durante os testes.**

**A primeira rede neural testada tinha como ativação nos neurônios a função ReLU. Tal modelo exibiu o melhor resultado durante a época 394 com um mse de 5.9162e-06. Já durante a fase de testes obteve um mse de 1.7216e-05, um pouco inferior ao melhor resultado na fase de treinamento, porém nada que configure sobreajuste. Dentre todas as redes neurais testadas, a rede com a função relu nas camadas intermediárias foi a que obteve os melhores resultados.**

**A segunda rede neural foi treinada utilizando-se a função de ativação Sigmoid nas camadas ocultas. Tal modelo exibiu o melhor resultado na fase de treinamento durante a época 399, obtendo um mse de 7.3958e-04. Durante o processo de testes, a rede em questão exibiu um mse de 7.6333e-04.**

**A terceira rede neural foi treinada com a função de ativação Tangente Hiperbólica - Tanh - dos neurônios das camadas intermediárias. Durante o processo de treinamento o melhor resultado foi obtido durante a época 394 exibindo um mse de 9.6290e-05 e durante a fase de testes o mse foi de 1.0117e-04, resultados ligeiramente superiores ao modelo com função Sigmoid.**

**O quarto e último modelo testado tinha nos neurônios das camadas ocultas as funções de ativação Sigmoid, Tanh e ReLU respectivamente. O melhor resultado obtido durante o treinamento do modelo foi um mse de 3.7251e-04 que ocorreu na última época - 400 e com um mse de 4.0191e-04 ao submeter a rede neural aos dados de testes.**

**Em resumo, a rede que obteve os melhores resultados tinha como a função ReLU como ativação das camadas intermediárias, a rede com o pior resultado foi a a rede em que se variou as funções de ativação em tais camadas. Os modelos com função Tanh e Sigmoid tiveram resultados intermediários, e embora não muito distantes do melhor modelo, porém que podem fazer a diferença dependendo do problema de previsão abordado.**


O que acontece se você mudar a função de otimização? Teste diferentes funções e descreva seus resultados, em especial as funções SDG e RMSprop


In [41]:
model = tf.keras.Sequential([    
    keras.Input(shape=(1,)),
    keras.layers.Dense(units = 10, activation = 'relu'),
    keras.layers.Dense(units = 10, activation = 'relu'),
    keras.layers.Dense(units = 10, activation = 'relu'),
    keras.layers.Dense(units = 1)
])
model.compile(optimizer="sgd", loss="mean_squared_error")
model.fit(x,y,epochs=400)

Epoch 1/400
4/4 [==============================] - 0s 926us/step - loss: 0.1675
Epoch 2/400
4/4 [==============================] - 0s 670us/step - loss: 0.0243
Epoch 3/400
4/4 [==============================] - 0s 2ms/step - loss: 0.0079
Epoch 4/400
4/4 [==============================] - 0s 2ms/step - loss: 0.0049
Epoch 5/400
4/4 [==============================] - 0s 1ms/step - loss: 0.0044
Epoch 6/400
4/4 [==============================] - 0s 2ms/step - loss: 0.0042
Epoch 7/400
4/4 [==============================] - 0s 2ms/step - loss: 0.0040
Epoch 8/400
4/4 [==============================] - 0s 2ms/step - loss: 0.0039
Epoch 9/400
4/4 [==============================] - 0s 3ms/step - loss: 0.0038
Epoch 10/400
4/4 [==============================] - 0s 1ms/step - loss: 0.0036
Epoch 11/400
4/4 [==============================] - 0s 2ms/step - loss: 0.0035
Epoch 12/400
4/4 [==============================] - 0s 2ms/step - loss: 0.0034
Epoch 13/400
4/4 [==============================] - 0s 2m

4/4 [==============================] - 0s 4ms/step - loss: 0.0015
Epoch 207/400
4/4 [==============================] - 0s 2ms/step - loss: 0.0014
Epoch 208/400
4/4 [==============================] - 0s 1ms/step - loss: 0.0014
Epoch 209/400
4/4 [==============================] - 0s 2ms/step - loss: 0.0014
Epoch 210/400
4/4 [==============================] - 0s 3ms/step - loss: 0.0014
Epoch 211/400
4/4 [==============================] - 0s 2ms/step - loss: 0.0014
Epoch 212/400
4/4 [==============================] - 0s 3ms/step - loss: 0.0014
Epoch 213/400
4/4 [==============================] - 0s 3ms/step - loss: 0.0014
Epoch 214/400
4/4 [==============================] - 0s 2ms/step - loss: 0.0014
Epoch 215/400
4/4 [==============================] - 0s 4ms/step - loss: 0.0015
Epoch 216/400
4/4 [==============================] - 0s 2ms/step - loss: 0.0014
Epoch 217/400
4/4 [==============================] - 0s 2ms/step - loss: 0.0014
Epoch 218/400
4/4 [==============================] - 0

In [42]:
model = tf.keras.Sequential([    
    keras.Input(shape=(1,)),
    keras.layers.Dense(units = 10, activation = 'relu'),
    keras.layers.Dense(units = 10, activation = 'relu'),
    keras.layers.Dense(units = 10, activation = 'relu'),
    keras.layers.Dense(units = 1)
])
model.compile(optimizer="RMSprop", loss="mean_squared_error")
model.fit(x,y,epochs=400)

Epoch 1/400
4/4 [==============================] - 0s 903us/step - loss: 0.4998
Epoch 2/400
4/4 [==============================] - 0s 843us/step - loss: 0.2212
Epoch 3/400
4/4 [==============================] - 0s 2ms/step - loss: 0.1360
Epoch 4/400
4/4 [==============================] - 0s 877us/step - loss: 0.0760
Epoch 5/400
4/4 [==============================] - 0s 2ms/step - loss: 0.0387
Epoch 6/400
4/4 [==============================] - 0s 3ms/step - loss: 0.0225
Epoch 7/400
4/4 [==============================] - 0s 3ms/step - loss: 0.0146
Epoch 8/400
4/4 [==============================] - 0s 2ms/step - loss: 0.0107
Epoch 9/400
4/4 [==============================] - 0s 2ms/step - loss: 0.0099
Epoch 10/400
4/4 [==============================] - 0s 2ms/step - loss: 0.0093
Epoch 11/400
4/4 [==============================] - 0s 2ms/step - loss: 0.0083
Epoch 12/400
4/4 [==============================] - 0s 2ms/step - loss: 0.0076
Epoch 13/400
4/4 [==============================] - 0s 

4/4 [==============================] - 0s 3ms/step - loss: 1.1520e-04
Epoch 201/400
4/4 [==============================] - 0s 3ms/step - loss: 4.2921e-04
Epoch 202/400
4/4 [==============================] - 0s 2ms/step - loss: 7.6159e-05
Epoch 203/400
4/4 [==============================] - 0s 2ms/step - loss: 2.9353e-04
Epoch 204/400
4/4 [==============================] - 0s 2ms/step - loss: 3.2715e-04
Epoch 205/400
4/4 [==============================] - 0s 2ms/step - loss: 3.9361e-04
Epoch 206/400
4/4 [==============================] - 0s 3ms/step - loss: 3.0387e-04
Epoch 207/400
4/4 [==============================] - 0s 2ms/step - loss: 6.5919e-05
Epoch 208/400
4/4 [==============================] - 0s 2ms/step - loss: 7.9946e-05
Epoch 209/400
4/4 [==============================] - 0s 2ms/step - loss: 1.2962e-04
Epoch 210/400
4/4 [==============================] - 0s 3ms/step - loss: 7.6372e-04
Epoch 211/400
4/4 [==============================] - 0s 2ms/step - loss: 8.0271e-05
Epoch 

4/4 [==============================] - 0s 2ms/step - loss: 2.6091e-05
Epoch 395/400
4/4 [==============================] - 0s 2ms/step - loss: 3.6337e-05
Epoch 396/400
4/4 [==============================] - 0s 1ms/step - loss: 5.0152e-04
Epoch 397/400
4/4 [==============================] - 0s 2ms/step - loss: 2.2564e-05
Epoch 398/400
4/4 [==============================] - 0s 1ms/step - loss: 9.6397e-05
Epoch 399/400
4/4 [==============================] - 0s 1ms/step - loss: 3.4702e-04
Epoch 400/400
4/4 [==============================] - 0s 2ms/step - loss: 9.3339e-05


In [46]:
model = tf.keras.Sequential([    
    keras.Input(shape=(1,)),
    keras.layers.Dense(units = 10, activation = 'relu'),
    keras.layers.Dense(units = 10, activation = 'relu'),
    keras.layers.Dense(units = 10, activation = 'relu'),
    keras.layers.Dense(units = 1)
])
model.compile(optimizer="Adadelta", loss="mean_squared_error")
model.fit(x,y,epochs=400)

Epoch 1/400
4/4 [==============================] - 0s 793us/step - loss: 0.0082
Epoch 2/400
4/4 [==============================] - 0s 696us/step - loss: 0.0082
Epoch 3/400
4/4 [==============================] - 0s 2ms/step - loss: 0.0082
Epoch 4/400
4/4 [==============================] - 0s 2ms/step - loss: 0.0082
Epoch 5/400
4/4 [==============================] - 0s 2ms/step - loss: 0.0082
Epoch 6/400
4/4 [==============================] - 0s 4ms/step - loss: 0.0082
Epoch 7/400
4/4 [==============================] - 0s 3ms/step - loss: 0.0081
Epoch 8/400
4/4 [==============================] - 0s 4ms/step - loss: 0.0081
Epoch 9/400
4/4 [==============================] - 0s 3ms/step - loss: 0.0081
Epoch 10/400
4/4 [==============================] - 0s 2ms/step - loss: 0.0081
Epoch 11/400
4/4 [==============================] - 0s 2ms/step - loss: 0.0081
Epoch 12/400
4/4 [==============================] - 0s 2ms/step - loss: 0.0081
Epoch 13/400
4/4 [==============================] - 0s 2m

4/4 [==============================] - 0s 3ms/step - loss: 0.0054
Epoch 207/400
4/4 [==============================] - 0s 2ms/step - loss: 0.0054
Epoch 208/400
4/4 [==============================] - 0s 1ms/step - loss: 0.0054
Epoch 209/400
4/4 [==============================] - 0s 2ms/step - loss: 0.0054
Epoch 210/400
4/4 [==============================] - 0s 2ms/step - loss: 0.0054
Epoch 211/400
4/4 [==============================] - 0s 2ms/step - loss: 0.0054
Epoch 212/400
4/4 [==============================] - 0s 2ms/step - loss: 0.0053
Epoch 213/400
4/4 [==============================] - 0s 2ms/step - loss: 0.0053
Epoch 214/400
4/4 [==============================] - 0s 2ms/step - loss: 0.0053
Epoch 215/400
4/4 [==============================] - 0s 2ms/step - loss: 0.0053
Epoch 216/400
4/4 [==============================] - 0s 2ms/step - loss: 0.0053
Epoch 217/400
4/4 [==============================] - 0s 2ms/step - loss: 0.0053
Epoch 218/400
4/4 [==============================] - 0

**<font color='red'> Sua resposta aqui </font>**

**Com o algoritmo de otimização SGD a rede neural atingiu o melhor resultado durante a época 87, um mse de 0.0014 e ficou estagnada neste mínimo até a época 400, ou seja, a rede se mostrou com pouca capacidade preditiva.**

**Em contrapartida, com algoritmo RMSprop, os resultados obtidos pelo modelo de rede neural foram superiores aos resultados exibidos pela rede com SGD, no entanto, embora não tenha ficado estagnada, os resultados obtidos em cada época de treinamento ficaram oscilando entre valores com magnitude de 10e-04 e 10e-05, caracterizando um pouco de ruído.**

**A rede com Adadelta não ficou estagnada nem ficou oscilando como as anteriores, o melhor resultado foi um mse de 0.0040 durante a época 400. Em comparação com os outros modelos, a rede utilizando este algoritmo demonstrou que precisa de mais épocas de treinamento para convergir para um mínimo local, já os outros algoritmos convergiram mais rapidamente.**

**O modelo que apresentou melhor desempenho durante a fase de treinamento, ou seja, sem os problemas observados nas redes utilizando os algoritmos supracitados, foi a rede neural que utilizou o algoritmo de otimização Adam.**

Volte a primeira parte desse notebook e troque a função de ativação da rede de uma camada (pérceptron) de sdg para adam, o que acontece?



**<font color='red'> Sua resposta aqui </font>**

**Ao substituir o algoritmo de otimização da rede Perceptron de SGD para Adam, os resultados exibidos pela função de custo aumentaram drasticamente, caracterizando assim uma piora significativa no desempenho do modelo em questão. Para se ter uma ideia, utilizando-se Adam obteve-se um mse de 2248.2883 durante e época 299, já com SGD  2.1232e-05 durante a época 300**

A avaliação de performance que realizamos foi apenas para pontos contidos no mesmo intervalo que o conjunto de treino, ou seja, foi apenas uma interpolação. Sem alterar sua rede repita o teste realizando uma extrapolação, com pontos fora do intervalo [0;10] e descreva seus resultados. O que aconteceu com a performance?

In [53]:
model = tf.keras.Sequential([    
    keras.Input(shape=(1,)),
    keras.layers.Dense(units = 10, activation = 'relu'),
    keras.layers.Dense(units = 10, activation = 'relu'),
    keras.layers.Dense(units = 10, activation = 'relu'),
    keras.layers.Dense(units = 1)
])
model.compile(optimizer="adam", loss="mean_squared_error")
model.fit(x,y,epochs=400)

Epoch 1/400
4/4 [==============================] - 0s 923us/step - loss: 0.6392
Epoch 2/400
4/4 [==============================] - 0s 694us/step - loss: 0.4866
Epoch 3/400
4/4 [==============================] - 0s 3ms/step - loss: 0.3524
Epoch 4/400
4/4 [==============================] - 0s 884us/step - loss: 0.2535
Epoch 5/400
4/4 [==============================] - 0s 3ms/step - loss: 0.1796
Epoch 6/400
4/4 [==============================] - 0s 2ms/step - loss: 0.1198
Epoch 7/400
4/4 [==============================] - 0s 2ms/step - loss: 0.0833
Epoch 8/400
4/4 [==============================] - 0s 2ms/step - loss: 0.0594
Epoch 9/400
4/4 [==============================] - 0s 3ms/step - loss: 0.0417
Epoch 10/400
4/4 [==============================] - 0s 2ms/step - loss: 0.0289
Epoch 11/400
4/4 [==============================] - 0s 2ms/step - loss: 0.0195
Epoch 12/400
4/4 [==============================] - 0s 3ms/step - loss: 0.0131
Epoch 13/400
4/4 [==============================] - 0s 

4/4 [==============================] - 0s 2ms/step - loss: 2.4446e-04
Epoch 198/400
4/4 [==============================] - 0s 4ms/step - loss: 2.4257e-04
Epoch 199/400
4/4 [==============================] - 0s 2ms/step - loss: 2.3952e-04
Epoch 200/400
4/4 [==============================] - 0s 2ms/step - loss: 2.3955e-04
Epoch 201/400
4/4 [==============================] - 0s 2ms/step - loss: 2.3713e-04
Epoch 202/400
4/4 [==============================] - 0s 2ms/step - loss: 2.3618e-04
Epoch 203/400
4/4 [==============================] - 0s 3ms/step - loss: 2.3515e-04
Epoch 204/400
4/4 [==============================] - 0s 3ms/step - loss: 2.3536e-04
Epoch 205/400
4/4 [==============================] - 0s 2ms/step - loss: 2.3466e-04
Epoch 206/400
4/4 [==============================] - 0s 2ms/step - loss: 2.3553e-04
Epoch 207/400
4/4 [==============================] - 0s 2ms/step - loss: 2.3396e-04
Epoch 208/400
4/4 [==============================] - 0s 3ms/step - loss: 2.3000e-04
Epoch 

4/4 [==============================] - 0s 2ms/step - loss: 9.9220e-05
Epoch 392/400
4/4 [==============================] - 0s 3ms/step - loss: 9.8166e-05
Epoch 393/400
4/4 [==============================] - 0s 3ms/step - loss: 9.8081e-05
Epoch 394/400
4/4 [==============================] - 0s 2ms/step - loss: 9.7899e-05
Epoch 395/400
4/4 [==============================] - 0s 3ms/step - loss: 9.6972e-05
Epoch 396/400
4/4 [==============================] - 0s 2ms/step - loss: 9.8240e-05
Epoch 397/400
4/4 [==============================] - 0s 2ms/step - loss: 9.8693e-05
Epoch 398/400
4/4 [==============================] - 0s 1ms/step - loss: 9.5812e-05
Epoch 399/400
4/4 [==============================] - 0s 2ms/step - loss: 9.6116e-05
Epoch 400/400
4/4 [==============================] - 0s 3ms/step - loss: 9.5028e-05


In [52]:
x_val = np.linspace(-10,20,50)
y_val = f2(x_val)

In [54]:
model.evaluate(x=x_val,y=y_val)

2/2 [==============================] - 0s 801us/step - loss: 0.3690


0.369022935628891

**<font color='red'> Sua resposta aqui </font>**

**O resultado obtido durante a fase de testes com valores fora do intervalo dos dados de entrada utilizados na fase de treinamento, se mostrou bastante inferior quando comparado com os resultados obtidos utilizando-se dados dentro da mesma faixa de valores da fase de treino. Isso nos mostra que o poder preditivo fica restrito a faixa de valores em que os dados apresentados para a rede durante a fase de treinamento se encontram.**

# Prevendo se vai chover na Austrália

## Pré processamento dos dados
Já definimos uma função basica para lhe ajudar a explorar os dados, você precisara escrever uma função para carregar os dados, uma de pré-processamento e outra para fazer a separação em teste e treino.

In [53]:
def visualize_data(data):
    """Gera graficos das distribuicoes das features e salva em disco"""

    ibm_pltt = ['#648FFF', '#785EF0', '#DC267F',
                '#FE6100', '#FFB000']  # Paleta colorblind-friendly
    plt.figure(figsize=(8, 6))

    # RainToday:
    sns.set()
    sns.set_palette(sns.color_palette([ibm_pltt[2], ibm_pltt[0]]))
    sns.countplot(data.RainToday)
    plt.xlabel('Choveu Hoje?')
    plt.ylabel('Contagem')
    plt.title("Valores de 'RainToday' para os dados pré-processados")
    plt.savefig('data_RainToday.png')
    plt.clf()

    # RainTomorrow:
    sns.set()
    sns.set_palette(sns.color_palette([ibm_pltt[3], ibm_pltt[1]]))
    sns.countplot(data.RainTomorrow)
    plt.xlabel('Choverá Amanhã?')
    plt.ylabel('Contagem')
    plt.title("Valores de 'RainTomorrow' para os dados pré-processados")
    plt.savefig('data_Rainomorrow.png')
    plt.clf()

    # Humidity3pm:
    sns.set()
    sns.distplot(data.Humidity3pm, color=ibm_pltt[0])
    plt.xlabel('Umidade às 3PM')
    plt.ylabel('Densidade normalizada')
    plt.title("Distribuição da variável 'Humidity3pm' para os dados pré-processados")
    plt.savefig('data_Humidity3pm.png')
    plt.clf()

    # Pressure9am:
    sns.set()
    sns.distplot(data.Pressure9am, color=ibm_pltt[4])
    plt.xlabel('Pressão atmosférica às 9AM')
    plt.ylabel('Densidade normalizada')
    plt.title("Distribuição da variável 'Pressure9amm' para os dados pré-processados")
    plt.savefig('data_Pressure9am.png')
    plt.clf()

    # Rainfall:
    sns.set()
    sns.distplot(data.Rainfall, color=ibm_pltt[1], bins=500, kde=False)
    plt.xlim(0, 10)
    plt.xlabel('Pluviosidade')
    plt.ylabel('Densidade normalizada')
    plt.title("Distribuição da variável 'Rainfall' para os dados pré-processados")
    plt.savefig('data_Rainfall.png')
    plt.clf()

    return


# <font color='blue'>Questão 3 </font>
Complete as funções aqui descritas seguindo a assinatura sugerida

Para separar em treino e teste, de uma olhada na função [train_test_split()](https://scikit-learn.org/stable/modules/generated/sklearn.model_selection.train_test_split.html) do Scikit-Learn

In [87]:
def load_data(data_path='data/weatherAUS.csv')-> pd.DataFrame:
    """Funcao que importa dados de um arquivo csv, usando pandas"""
    
    raw_data = pd.read_csv(data_path)

    return raw_data

def pre_processing(raw_data:pd.DataFrame)-> pd.DataFrame:
    """Funcao que filtra e limpa os dados meteorologicos para o treinamento"""
    
    columns = ['Rainfall', 'Humidity3pm', 'Pressure9am', 'RainToday', 'RainTomorrow']
    processed_data = raw_data[columns].copy()
    processed_data.dropna(inplace=True)
    processed_data['RainToday'] = processed_data['RainToday'].apply(lambda answer : 
                                                                    1 if answer.lower() == 'yes'
                                                                    else 0)
    processed_data['RainTomorrow'] = processed_data['RainTomorrow'].apply(lambda answer : 
                                                                          1 if answer.lower() == 'yes'
                                                                          else 0)
    return processed_data


def split_data(data:pd.DataFrame, val_size= 0.2)-> np.array:
    """Funcao que separa seus dados em teste e treino conforme a proporcao val_size"""
    
    from sklearn.preprocessing import MinMaxScaler
    
    y = data['RainTomorrow'].values
    data.drop(['RainTomorrow'], axis=1, inplace=True)
    X = data.values
    sc = MinMaxScaler()  # Normalização dos dados
    sc.fit(X[:40000])
    X = sc.transform(X)
    X_train, X_val, y_train, y_val = train_test_split(X, y, test_size=val_size)
   
    return X_train, X_val, y_train, y_val


In [88]:
df = load_data()
visualize_data(df.dropna())
df = pre_processing(df)
x_train, x_val, y_train, y_val = split_data(df,)

<Figure size 576x432 with 0 Axes>

# <font color='blue'>Questão 4 </font>
Agora que você ja ganhou uma familiaridade com a API Keras, escreva sozinho do começo ao fim um modelo que ira dizer se amanhã vai chover ou não e avalie sua performance.


# Modelo 1

3 Camadas ocultas

512 neurônios primeira camada

256 neurônios segunda camada

128 neurônios terceira camada

Função de ativação camadas ocultas: ReLU

Função de ativação camada de saída: Sigmoid

Inicialização dos pesos: Glorot Uniform

Loss: Binary Crossentropy

Dropout de 10%

Otimizador: Adam

Batch Size: 64

Épocas: 50

In [96]:
def make_model(input_size):
    model = tf.keras.Sequential()
    model.add(tf.keras.layers.Dense(units=512, kernel_initializer='glorot_uniform', activation='relu', input_dim=input_size))
    model.add(tf.keras.layers.Dense(units=256, kernel_initializer='glorot_uniform', activation='relu'))
    model.add(tf.keras.layers.Dense(units=128, kernel_initializer='glorot_uniform', activation='relu'))
    model.add(tf.keras.layers.Dropout(0.1))
    model.add(tf.keras.layers.Dense(units=1, kernel_initializer='glorot_uniform', activation='sigmoid'))
    model.compile(optimizer = 'adam', loss='binary_crossentropy', metrics=['accuracy'])
    return model

In [97]:
model = make_model(x_train.shape[1])
history = model.fit(x_train, y_train, batch_size=64, epochs=50)

Epoch 1/50
1559/1559 [==============================] - 4s 3ms/step - loss: 0.3858 - accuracy: 0.8360
Epoch 2/50
1559/1559 [==============================] - 5s 3ms/step - loss: 0.3808 - accuracy: 0.8388
Epoch 3/50
1559/1559 [==============================] - 5s 3ms/step - loss: 0.3794 - accuracy: 0.8398
Epoch 4/50
1559/1559 [==============================] - 5s 3ms/step - loss: 0.3794 - accuracy: 0.8396
Epoch 5/50
1559/1559 [==============================] - 5s 3ms/step - loss: 0.3785 - accuracy: 0.8392
Epoch 6/50
1559/1559 [==============================] - 5s 3ms/step - loss: 0.3785 - accuracy: 0.8397
Epoch 7/50
1559/1559 [==============================] - 5s 3ms/step - loss: 0.3779 - accuracy: 0.8400
Epoch 8/50
1559/1559 [==============================] - 6s 4ms/step - loss: 0.3779 - accuracy: 0.8397
Epoch 9/50
1559/1559 [==============================] - 6s 4ms/step - loss: 0.3775 - accuracy: 0.8390
Epoch 10/50
1559/1559 [==============================] - 6s 4ms/step - loss: 0.377

In [98]:
model.evaluate(x=x_val,y=y_val)

780/780 [==============================] - 1s 997us/step - loss: 0.3829 - accuracy: 0.8362


[0.38285815715789795, 0.8362338542938232]

# Modelo 2

3 Camadas ocultas

512 neurônios primeira camada

256 neurônios segunda camada

128 neurônios terceira camada

Função de ativação camadas ocultas: Tanh

Função de ativação camada de saída: Sigmoid

Inicialização dos pesos: Glorot Uniform

Loss: Binary Crossentropy

Dropout de 10%

Otimizador: Adam

Batch Size: 64

Épocas: 50


In [99]:
def make_model(input_size):
    model = tf.keras.Sequential()
    model.add(tf.keras.layers.Dense(units=512, kernel_initializer='glorot_uniform', activation='tanh', input_dim=input_size))
    model.add(tf.keras.layers.Dense(units=256, kernel_initializer='glorot_uniform', activation='tanh'))
    model.add(tf.keras.layers.Dense(units=128, kernel_initializer='glorot_uniform', activation='tanh'))
    model.add(tf.keras.layers.Dropout(0.1))
    model.add(tf.keras.layers.Dense(units=1, kernel_initializer='glorot_uniform', activation='sigmoid'))
    model.compile(optimizer = 'adam', loss='binary_crossentropy', metrics=['accuracy'])
    return model

In [100]:
model = make_model(x_train.shape[1])
history = model.fit(x_train, y_train, batch_size=64, epochs=50)

Epoch 1/50
1559/1559 [==============================] - 4s 3ms/step - loss: 0.3874 - accuracy: 0.8365
Epoch 2/50
1559/1559 [==============================] - 5s 3ms/step - loss: 0.3836 - accuracy: 0.8384
Epoch 3/50
1559/1559 [==============================] - 4s 3ms/step - loss: 0.3821 - accuracy: 0.8381
Epoch 4/50
1559/1559 [==============================] - 4s 3ms/step - loss: 0.3817 - accuracy: 0.8382
Epoch 5/50
1559/1559 [==============================] - 4s 3ms/step - loss: 0.3812 - accuracy: 0.8392
Epoch 6/50
1559/1559 [==============================] - 4s 3ms/step - loss: 0.3809 - accuracy: 0.8380
Epoch 7/50
1559/1559 [==============================] - 4s 3ms/step - loss: 0.3811 - accuracy: 0.8385
Epoch 8/50
1559/1559 [==============================] - 4s 3ms/step - loss: 0.3803 - accuracy: 0.8382
Epoch 9/50
1559/1559 [==============================] - 4s 3ms/step - loss: 0.3805 - accuracy: 0.8384
Epoch 10/50
1559/1559 [==============================] - 5s 3ms/step - loss: 0.380

In [102]:
model.evaluate(x=x_val,y=y_val)

780/780 [==============================] - 1s 1ms/step - loss: 0.3844 - accuracy: 0.8355


[0.38439393043518066, 0.8354719877243042]

# Modelo 3

1 Camada oculta

128 neurônios na primeira camada

Função de ativação camadas ocultas: ReLU

Função de ativação camada de saída: Sigmoid

Inicialização dos pesos: Glorot Uniform

Loss: Binary Crossentropy

Dropout de 10%

Otimizador: Adam

Batch Size: 64

Épocas: 50

In [110]:
def make_model(input_size):
    model = tf.keras.Sequential()
    model.add(tf.keras.layers.Dense(units=128, activation='relu', kernel_initializer='glorot_uniform', input_dim=input_size))
    model.add(tf.keras.layers.Dropout(0.1))
    model.add(tf.keras.layers.Dense(units=1, kernel_initializer='glorot_uniform', activation='sigmoid'))
    model.compile(optimizer = 'adam', loss='binary_crossentropy', metrics=['accuracy'])
    return model

In [111]:
model = make_model(x_train.shape[1])
history = model.fit(x_train, y_train, batch_size=64, epochs=50)

Epoch 1/50
1559/1559 [==============================] - 2s 1ms/step - loss: 0.4072 - accuracy: 0.8262
Epoch 2/50
1559/1559 [==============================] - 2s 1ms/step - loss: 0.3815 - accuracy: 0.8385
Epoch 3/50
1559/1559 [==============================] - 2s 1ms/step - loss: 0.3803 - accuracy: 0.8389
Epoch 4/50
1559/1559 [==============================] - 3s 2ms/step - loss: 0.3799 - accuracy: 0.8388
Epoch 5/50
1559/1559 [==============================] - 3s 2ms/step - loss: 0.3795 - accuracy: 0.8387
Epoch 6/50
1559/1559 [==============================] - 2s 1ms/step - loss: 0.3788 - accuracy: 0.8389
Epoch 7/50
1559/1559 [==============================] - 2s 1ms/step - loss: 0.3783 - accuracy: 0.8394
Epoch 8/50
1559/1559 [==============================] - 2s 1ms/step - loss: 0.3778 - accuracy: 0.8398
Epoch 9/50
1559/1559 [==============================] - 2s 1ms/step - loss: 0.3773 - accuracy: 0.8403
Epoch 10/50
1559/1559 [==============================] - 2s 1ms/step - loss: 0.377

In [112]:
model.evaluate(x=x_val,y=y_val)

780/780 [==============================] - 1s 887us/step - loss: 0.3826 - accuracy: 0.8368


[0.3825942277908325, 0.8367551565170288]

# Modelo 4

1 Camada oculta

128 neurônios na primeira camada

Função de ativação camadas ocultas: Tanh

Função de ativação camada de saída: Sigmoid

Inicialização dos pesos: Glorot Uniform

Loss: Binary Crossentropy

Dropout de 10%

Otimizador: Adam

Batch Size: 64

Épocas: 50

In [118]:
def make_model(input_size):
    model = tf.keras.Sequential()
    model.add(tf.keras.layers.Dense(units=128, activation='sigmoid', kernel_initializer='glorot_uniform', input_dim=input_size))
    model.add(tf.keras.layers.Dense(units=1, kernel_initializer='glorot_uniform', activation='sigmoid'))
    model.compile(optimizer = 'adam', loss='binary_crossentropy', metrics=['accuracy'])
    return model

In [119]:
model = make_model(x_train.shape[1])
history = model.fit(x_train, y_train, batch_size=64, epochs=50)

Epoch 1/50
1559/1559 [==============================] - 2s 964us/step - loss: 0.4411 - accuracy: 0.8074
Epoch 2/50
1559/1559 [==============================] - 2s 1ms/step - loss: 0.3858 - accuracy: 0.8382
Epoch 3/50
1559/1559 [==============================] - 2s 990us/step - loss: 0.3825 - accuracy: 0.8390
Epoch 4/50
1559/1559 [==============================] - 2s 986us/step - loss: 0.3824 - accuracy: 0.8384
Epoch 5/50
1559/1559 [==============================] - 2s 967us/step - loss: 0.3822 - accuracy: 0.8376
Epoch 6/50
1559/1559 [==============================] - 1s 899us/step - loss: 0.3821 - accuracy: 0.8383
Epoch 7/50
1559/1559 [==============================] - 2s 984us/step - loss: 0.3820 - accuracy: 0.8383
Epoch 8/50
1559/1559 [==============================] - 1s 923us/step - loss: 0.3821 - accuracy: 0.8381
Epoch 9/50
1559/1559 [==============================] - 2s 1ms/step - loss: 0.3820 - accuracy: 0.8386
Epoch 10/50
1559/1559 [==============================] - 2s 1ms/step

In [120]:
model.evaluate(x=x_val,y=y_val)

780/780 [==============================] - 1s 661us/step - loss: 0.3896 - accuracy: 0.8337


[0.389588326215744, 0.8337075710296631]

# Modelo 5

2 Camadas ocultas

128 neurônios na primeira camada

128 neurônios na segunda camada

Função de ativação camadas ocultas: ReLU

Função de ativação camada de saída: Sigmoid

Inicialização dos pesos: Glorot Uniform

Loss: Binary Crossentropy

Dropout de 10%

Otimizador: Adam

Batch Size: 64

Épocas: 50

In [121]:
def make_model(input_size):
    model = tf.keras.Sequential()
    model.add(tf.keras.layers.Dense(units=128, activation='relu', kernel_initializer='glorot_uniform', input_dim=input_size))
    model.add(tf.keras.layers.Dense(units=128, activation='relu', kernel_initializer='glorot_uniform'))
    model.add(tf.keras.layers.Dropout(0.1))
    model.add(tf.keras.layers.Dense(units=1, kernel_initializer='glorot_uniform', activation='sigmoid'))
    model.compile(optimizer = 'adam', loss='binary_crossentropy', metrics=['accuracy'])
    return model

In [122]:
model = make_model(x_train.shape[1])
history = model.fit(x_train, y_train, batch_size=64, epochs=50)

Epoch 1/50
1559/1559 [==============================] - 2s 1ms/step - loss: 0.3884 - accuracy: 0.8354
Epoch 2/50
1559/1559 [==============================] - 2s 1ms/step - loss: 0.3810 - accuracy: 0.8378
Epoch 3/50
1559/1559 [==============================] - 2s 1ms/step - loss: 0.3792 - accuracy: 0.8388
Epoch 4/50
1559/1559 [==============================] - 2s 1ms/step - loss: 0.3784 - accuracy: 0.8392
Epoch 5/50
1559/1559 [==============================] - 2s 1ms/step - loss: 0.3781 - accuracy: 0.8395
Epoch 6/50
1559/1559 [==============================] - 2s 1ms/step - loss: 0.3777 - accuracy: 0.8397
Epoch 7/50
1559/1559 [==============================] - 2s 1ms/step - loss: 0.3774 - accuracy: 0.8395
Epoch 8/50
1559/1559 [==============================] - 2s 1ms/step - loss: 0.3775 - accuracy: 0.8399
Epoch 9/50
1559/1559 [==============================] - 2s 1ms/step - loss: 0.3771 - accuracy: 0.8398
Epoch 10/50
1559/1559 [==============================] - 2s 1ms/step - loss: 0.377

In [123]:
model.evaluate(x=x_val,y=y_val)

780/780 [==============================] - 1s 730us/step - loss: 0.3827 - accuracy: 0.8365


[0.38269472122192383, 0.8365145325660706]

**Cinco modelos de redes neurais foram treinados e testados, variando-se a quantidade de camadas ocultas, número de neurônios e funções de ativação. O modelo que obteve o melhor desempenho foi o modelo 3, que atingiu uma acurácia de 84,09% durante a época 29 e 84,68% no conjunto de testes. Tal modelo possuí uma camada oculta com 128 neurônios com função de ativação ReLU, 10% de Dropout após tal camada e um neurônio na camada de saída com função de ativação Sigmoid.**

**Com os resultados obtidos através dos modelos desenvolvidos, pode-se observar que os modelos não se mostraram representativos, pois, todos eles apresentaram uma acurácia de 82% logo de início e ficaram estagnados entre 83% e 84% sem mostrarem nenhum sinal de melhora durante o decorrer to treinamento, apenas se observou flutuações nas últimas casas decimais.**

**Embora 84% de acurácia seja um resultado razoável, apenas esse resultado não pode ser levado em consideração levando-se em conta que as classes que estamos a prever estão desbalanceadas. Para lidar com essa situação existem várias abordagens que poderiam ser tomadas, como a utilização de métricas adequadas para dados desbalanceados e a ponderação das classes com com pesos. Em relação a obter modelos mais representativos e com melhor acurácia é necessário fazer uma análise mais minuciosa dos dados, de modo a levantar novas hipóteses, e assim, testar diferentes combinações de *features* e aplicar outras técnicas, e testar com outros modelos.**